raiden_contracts/contracts/services/UserDeposit.sol

```diff
@@ -0,0 +1,171 @@
+pragma solidity ^0.5.2;
+
+import "raiden/Token.sol";
+import "raiden/Utils.sol";
+
+contract UserDeposit is Utils {
+    string constant public contract_version = "0.5.0";
+    uint constant public withdraw_delay = 100;  // time before withdraw is allowed in blocks
+
+    // Token to be used for the deposit
+    Token public token;
+
+    // Trusted contract (can execute `transfer`)
+    address public msc_address;
+
+    // Keep track of balances
+    mapping(address => uint256) public balances;
+    mapping(address => WithdrawPlan) public withdraw_plans;
+
+    /*
+     *  Structs
+     */
+    struct WithdrawPlan {
+        uint256 amount;
+        uint256 withdraw_block;  // earliest block at which withdraw is allowed
+    }
+
+    /*
+     *  Events
+     */
+
+    event BalanceReduced(address indexed owner, uint newBalance);
+    event WithdrawPlanned(address indexed withdrawer, uint plannedBalance);
+
+    /*
+     *  Modifiers
+     */
+
+    modifier canTransfer() {
+        require(msg.sender == msc_address);
+        _;
+    }
+
+    /*
+     *  Constructor
+     */
+
+    /// @notice Set the default values for the smart contract
+    /// @param _token_address The address of the token to use for rewards
+    constructor(address _token_address)
+        public
+    {
+        // check token contract
+        require(_token_address != address(0x0));
+        require(contractExists(_token_address));
+        token = Token(_token_address);
+        require(token.totalSupply() > 0); // Check if the contract is indeed a token contract
+    }
+
+    /// @notice Specify trusted contracts. This has to be done outside of the
+    /// constructor to avoid cyclic dependencies.
+    /// @param _msc_address Address of the MonitoringService contract
+    function init(address _msc_address)
+        external
+    {
+        // prevent changes of trusted contract after initialization
+        require(msc_address == address(0x0));
+
+        // check monitoring service contract
+        require(_msc_address != address(0x0));
+        require(contractExists(_msc_address));
+        msc_address = _msc_address;
+    }
+
+    /// @notice Deposit tokens. Idempotent function that sets the
+    /// total_deposit of tokens of the beneficiary.
+    /// Can be called by anyone several times and on behalf of other accounts
+    /// @param beneficiary The account benefiting from the deposit
+    /// @param total_deposit The amount tokens in the user's deposit after the call
+    function deposit(address beneficiary, uint256 total_deposit)
+        external
+    {
+        require(total_deposit > balances[beneficiary]);
+
+        // Calculate the actual amount of tokens that will be transferred
+        uint256 added_deposit = total_deposit - balances[beneficiary];
+
+        require(token.transferFrom(msg.sender, address(this), added_deposit));
```

Comment:
Hmm, with my comment from https://github.com/raiden-network/raiden-contracts/pull/448/files#r250609427 `(meaning, we always have state checks before the transfer, as you also have here)` I actually meant `state changes before the transfer` - so, exactly what @pirapira is saying above. Did I say something that prompted you to change the order?

Reply:
While changing that I was thinking about first executing the negative effects for the user so that he didn't gain any balance when weird things happen inside of `transferFrom`. But this is not helpful because everything is done in a transactional context anyway and it leads to the problem you both pointed to. So my temporary confusion is cleared up now and I will fix this, again.

Reply:
Yeah this environment is a bit different from, say, OS programming (where you should often care about the intermediate states being visible from outside).